optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,543 @@
+                if np.any(np.isnan(np.array(_constraints))):
+                    raise ValueError("NaN is not acceptable as constraint value.")
+
+        dominated_count: DefaultDict[int, int] = defaultdict(int)
+        dominates_list = defaultdict(list)
+
+        dominates = _dominates if self._constraints_func is None else _constrained_dominates
+
+        for p, q in itertools.combinations(population, 2):
+            if dominates(p, q, directions):
+                dominates_list[p.number].append(q.number)
+                dominated_count[q.number] += 1
+            elif dominates(q, p, directions):
+                dominates_list[q.number].append(p.number)
+                dominated_count[p.number] += 1
+
+        population_per_rank = []
+        while population:
+            non_dominated_population = []
+            i = 0
+            while i < len(population):
+                if dominated_count[population[i].number] == 0:
+                    individual = population[i]
+                    if i == len(population) - 1:
+                        population.pop()
+                    else:
+                        population[i] = population.pop()
+                    non_dominated_population.append(individual)
+                else:
+                    i += 1
+
+            for x in non_dominated_population:
+                for y in dominates_list[x.number]:
+                    dominated_count[y] -= 1
+
+            assert non_dominated_population
+            population_per_rank.append(non_dominated_population)
+
+        return population_per_rank
+
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        assert state in [TrialState.COMPLETE, TrialState.FAIL, TrialState.PRUNED]
+        if self._constraints_func is not None:
+            _process_constraints_after_trial(self._constraints_func, study, trial, state)
+        self._random_sampler.after_trial(study, trial, state, values)
+
+
+def multi_choose(n: int, k: int) -> int:
+    return math.factorial(n + k - 1) // math.factorial(k) // math.factorial(n - 1)
+
+
+def generate_default_reference_point(
+    objective_dimension: int, dividing_parameter: int = 3
+) -> np.ndarray:
+    """Generates default reference points which are `uniformly` spread on a hyperplane."""
+    reference_points = np.zeros(
+        (
+            multi_choose(objective_dimension, dividing_parameter),
+            objective_dimension,
+        )
+    )
+    for i, comb in enumerate(
+        itertools.combinations_with_replacement(range(objective_dimension), dividing_parameter)
+    ):
+        for j in comb:
+            reference_points[i, j] += 1.0
+    return reference_points
+
+
+def _normalize(
+    elite_population: List[FrozenTrial],
+    population: List[FrozenTrial],
+    weights: Optional[np.ndarray] = None,
+    method: str = "original",
```

Comment:
Logics for this `weights` and `method` are dead codes (not used and not tested). Should we leave these arguments on this PR? (I think it's OK to leave them as they are, as long as we work on propagating them to the constructor ASAP after this PR.)

Reply:
I romoved the options for now.